features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher-new/source/src/app/components/Apis/Details/Configuration/components/AccessControl.jsx

```diff
@@ -38,8 +44,79 @@ import Select from '@material-ui/core/Select';
  * @returns
  */
 export default function AccessControl(props) {
+    const [roleValidity, setRoleValidity] = useState(true);
+    const [userRoleValidity, setUserRoleValidity] = useState(true);
     const { api, configDispatcher } = props;
     const isRestricted = api.accessControl === 'RESTRICTED';
+    const [invalidRoles, setInvalidRoles] = useState([]);
+    const handleRoleAddition = (systemValidation, userValidation, role) => {
+        systemValidation.then((resp) => {
+            if (resp) {
+                setRoleValidity(true);
+                userValidation.then((response) => {
+                    if (response) {
+                        setUserRoleValidity(true);
+                        configDispatcher({
+                            action: 'accessControlRoles',
+                            value: api.accessControlRoles.length === 0 ? role : (api.accessControlRoles + ',' + role),
+                        });
+                    } else {
+                        setUserRoleValidity(false);
+                        setInvalidRoles([...invalidRoles, role]);
+                    }
+                }).catch((error) => {
+                    console.error('Error when validating user roles ' + error);
+                });
+            } else {
+                setRoleValidity(false);
+                setInvalidRoles([...invalidRoles, role]);
+            }
+        }).catch((error) => {
+            console.error('Error when validating roles ' + error);
+        });
+    };
+
+    const handleRoleDeletion = (role) => {
+        const validRoles = api.accessControlRoles;
+        const index = validRoles.indexOf(role);
+        if (index > -1) {
+            validRoles.splice(index, 1);
+        } else {
+            invalidRoles.splice(invalidRoles.indexOf(role), 1);
+            setInvalidRoles(invalidRoles);
+            if (invalidRoles.length === 0) {
+                setRoleValidity(true);
+                setUserRoleValidity(true);
+            }
+        }
+        configDispatcher({ action: 'accessControlRoles', value: validRoles.length === 0 ? '' : validRoles.join(',') });
```

Comment:
I think this is redundant work here, We are joining the values here and splitting the comma-separated  value in the reducer, I think we should always keep the values in an array structure.

Reply:
Fixed in f20d89587c00c4aadaf582e15f4f469743872571